### PR TITLE
Fix docstring for `value_and_pullback_function`

### PR DESCRIPTION
**Pull request by gdalle**
Fix the docstring of `value_and_pullback_function` to solve #124 and #119

Clarify that 
- `value_and_pushforward_function` means (value and pushforward) function
- `value_and_pullback_function` means value and (pullback function)

Name some anonymous functions in the process to make the code more readable

### CODE REVIEW

File: src/AbstractDifferentiation.jl [hunk 1]
```diff
@@ -226,44 +226,48 @@ The pushfoward function `pf` accepts as input a `Tuple` of tangents, one for eac
...
-                    newxs = xs .+ ds .* xds
-                    return f(newxs...)
-                else
-                    newx = only(xs) + ds * only(xds)
-                    return f(newx)
-                end
-            end,
-            _zero.(xs, ds)...,
-        )
+    function pf(ds)
```
Comment (devmotion) [inline src/AbstractDifferentiation.jl:229]: Generally I would have preferred to not include such changes in the PR as they are not related to the docstring fixes and make the PR more difficult to review.

File: src/AbstractDifferentiation.jl [hunk 2]
```diff
@@ -291,21 +295,24 @@ The pullback function `pb` accepts as input a `Tuple` of cotangents, one for eac
 If `f` has a single output, `pb` can also accept a single input instead of a 1-tuple.
 """
 function pullback_function(ab::AbstractBackend, f, xs...)
-    _, pbf = value_and_pullback_function(ab, f, xs...)
-    return pbf
+    _, pb = value_and_pullback_function(ab, f, xs...)
```
Comment (devmotion) [inline src/AbstractDifferentiation.jl:298]: What's the motivation for this change? `pbf` seemed to refer to "pullback function" I assume which seems to be the correct interpretation?

File: src/AbstractDifferentiation.jl (hunk 1 above)
Comment (gdalle) [reply src/AbstractDifferentiation.jl:229]: Generally I would agree but I included this refactoring because it makes this part of the code clearer. I was the one who wrote all the docstrings the first time around, and I screwed up on `value_and_pullback_function` partly because the code was a bit obscure

File: src/AbstractDifferentiation.jl (hunk 2 above)
Comment (gdalle) [reply src/AbstractDifferentiation.jl:298]: I think the terms "pullback" and "pullback function" are both valid, see eg. the JAX docs. My rationale for this was to fight ambiguities, because the function returning pb is already called pullback_function. So the idea is that the pullback function returns a pullback, which is itself a function

File: src/AbstractDifferentiation.jl (hunk 1 above)
Comment (devmotion) [reply src/AbstractDifferentiation.jl:229]: Yeah, I'm fine with it being included in this PR but wanted to mention it nevertheless as a general consideration.

File: src/AbstractDifferentiation.jl (hunk 2 above)
Comment (devmotion) [reply src/AbstractDifferentiation.jl:298]: My feeling was that the change here actually made it less clear - `gradient` etc. return gradients, and similarly `pullback_function` returns a "pullback function" for which `pbf` seemed to be an appropriate abbreviation.

File: src/AbstractDifferentiation.jl (hunk 1 above)
Comment (gdalle) [reply src/AbstractDifferentiation.jl:229]: Then we're in agreement :)

File: src/AbstractDifferentiation.jl (hunk 2 above)
Comment (gdalle) [reply src/AbstractDifferentiation.jl:298]: Then we should also turn `pf` into `pff` on the pushforward side (another reason for my change). I don't have strong opinions either way, it just feels weird to me that `pullback_function` returns a `pbf` which is itself short for... pullback function

Comment (gdalle) [reply src/AbstractDifferentiation.jl]: fixed in my latest commit

**Review by devmotion — APPROVED** [review]
